File: windows/src/HoverPocket.Shell/Capabilities/TodayFocusTextAdapter.cs
using System.Text;
using System.Text.Json;
using System.Globalization;

namespace HoverPocket.Shell.Capabilities;

internal sealed record TodayFocusCalendarEvent(
    string EventRef,
    string SafeTitle,
    DateTimeOffset Start,
    DateTimeOffset End);

internal sealed record TodayFocusDraft(
    CapabilityExecutionPlan Plan,
    CapabilityBrokerPreparation Preparation);

internal static class TodayFocusApprovalText
{
    private static readonly HashSet<int> BidirectionalControls =
    [
        0x061C, 0x200E, 0x200F,
        0x202A, 0x202B, 0x202C, 0x202D, 0x202E,
        0x2066, 0x2067, 0x2068, 0x2069
    ];

    public static string Sanitize(string value)
    {
        var builder = new StringBuilder();
        var pendingSpace = false;
        foreach (var rune in value.EnumerateRunes())
        {
            var category = Rune.GetUnicodeCategory(rune);
            var disallowed = category is UnicodeCategory.Control
                or UnicodeCategory.Format
                or UnicodeCategory.LineSeparator
                or UnicodeCategory.ParagraphSeparator
                || BidirectionalControls.Contains(rune.Value);
            if (disallowed || Rune.IsWhiteSpace(rune))
            {
                pendingSpace = builder.Length > 0;
                continue;
            }
            if (pendingSpace)
            {
                builder.Append(' ');
                pendingSpace = false;
            }
            builder.Append(rune.ToString());
        }
        var normalized = builder.ToString().Trim();
        return string.IsNullOrEmpty(normalized)
            ? "予定名なし"
            : CapabilityJson.TruncateString(normalized, 120);
    }
}

internal sealed class TodayFocusTextAdapter(CapabilityBroker broker)
{
    private readonly CapabilityBroker _broker = broker;

    public async Task<IReadOnlyList<TodayFocusCalendarEvent>> ListTodayAsync(
        string timezoneId,
        CapabilityPrincipal principal,
        CapabilityPermissionSet permissions,
        DateTimeOffset now,
        CancellationToken cancellationToken = default)
    {
        var nonce = Guid.NewGuid().ToString("D");
        var plan = new CapabilityExecutionPlan(
            $"today-focus-read:{nonce}",
            now,
            CapabilityOrigin.Text,
            principal,
            null,
            [
                new CapabilityPlanStep(
                    "listCalendar",
                    CapabilityIds.CalendarList,
                    CapabilityJson.From(new { range = "today", timezone = timezoneId }),
                    $"today-focus-read.{nonce}",
                    [])
            ],
            new HashSet<string>(["calendar.events.read"], StringComparer.Ordinal));
        var preparation = _broker.Prepare(plan, permissions, now);
        if (preparation.ApprovalRequest is not null)
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "read_approval");
        }
        var receipt = await _broker.ExecuteAsync(plan, permissions, null, now, cancellationToken);
        if (receipt.Status != CapabilityReceiptStatus.Succeeded
            || receipt.Steps.FirstOrDefault()?.Output is not { } output
            || !output.TryGetProperty("events", out var events)
            || events.ValueKind != JsonValueKind.Array)
        {
            throw new CapabilityBrokerException("CAPABILITY_UNAVAILABLE", CapabilityIds.CalendarList.Id);
        }
        return events.EnumerateArray().Select(item =>
        {
            if (!item.TryGetProperty("eventRef", out var eventRef)
                || !item.TryGetProperty("safeTitle", out var safeTitle)
                || !item.TryGetProperty("start", out var start)
                || !item.TryGetProperty("end", out var end)
                || !DateTimeOffset.TryParse(start.GetString(), out var startDate)
                || !DateTimeOffset.TryParse(end.GetString(), out var endDate))
            {
                throw new CapabilityBrokerException("CAPABILITY_READBACK_MISMATCH", "calendar.events");
            }
            return new TodayFocusCalendarEvent(
                eventRef.GetString() ?? string.Empty,
                safeTitle.GetString() ?? string.Empty,
                startDate,
                endDate);
        }).ToArray();
    }

    public TodayFocusDraft PrepareFocus(
        TodayFocusCalendarEvent selectedEvent,
        int durationSeconds,
        string purpose,
        CapabilityPrincipal principal,
        CapabilityPermissionSet permissions,
        DateTimeOffset now,
        TimeZoneInfo? timeZone = null)
    {
        if (durationSeconds is < 1 or > 86_400
            || string.IsNullOrEmpty(purpose)
            || purpose.EnumerateRunes().Count() > 10_000)
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "today_focus_input");
        }
        var nonce = Guid.NewGuid().ToString("D");
        var plan = new CapabilityExecutionPlan(
            $"today-focus-write:{nonce}",
            now,
            CapabilityOrigin.Text,
            principal,
            null,
            [
                new CapabilityPlanStep(
                    "startTimer",
                    CapabilityIds.TimerStart,
                    CapabilityJson.From(new
                    {
                        durationSeconds,
                        title = CapabilityJson.TruncateString(selectedEvent.SafeTitle, 80),
                        sourceRef = selectedEvent.EventRef
                    }),
                    $"today-focus-timer.{nonce}",
                    []),
                new CapabilityPlanStep(
                    "savePurpose",
                    CapabilityIds.StickyUpsert,
                    CapabilityJson.From(new
                    {
                        stableKey = $"today-focus:{TimeZoneInfo.ConvertTime(now, timeZone ?? TimeZoneInfo.Local):yyyy-MM-dd}",
                        title = "今日の目的",
                        body = purpose,
                        color = "yellow"
                    }),
                    $"today-focus-sticky.{nonce}",
                    ["startTimer"])
            ],
            new HashSet<string>(["sticky.write", "timer.write"], StringComparer.Ordinal));
        return new TodayFocusDraft(plan, _broker.Prepare(plan, permissions, now));
    }

    public async Task<CapabilityWorkflowReceipt> ApproveAndExecuteAsync(
        TodayFocusDraft draft,
        CapabilityPermissionSet permissions,
        DateTimeOffset now,
        CancellationToken cancellationToken = default)
    {
        var request = draft.Preparation.ApprovalRequest
            ?? throw new CapabilityBrokerException("CAPABILITY_APPROVAL_REQUIRED", draft.Plan.Id);
        var grant = _broker.DecideApproval(
            request.Id,
            draft.Preparation.PlanDigest,
            CapabilityApprovalDecision.Approve,
            now);
        return await _broker.ExecuteAsync(draft.Plan, permissions, grant, now, cancellationToken);
    }
}

File: Sources/HoverPocket/Capabilities/TodayFocusTextAdapter.swift
import Foundation

struct TodayFocusCalendarEvent: Equatable, Sendable {
    let eventRef: String
    let safeTitle: String
    let start: Date
    let end: Date
}

struct TodayFocusDraft: Equatable, Sendable {
    let plan: CapabilityExecutionPlan
    let preparation: CapabilityBrokerPreparation
    let approvalText: String
}

enum TodayFocusApprovalText {
    private static let bidirectionalControls: Set<UInt32> = [
        0x061C, 0x200E, 0x200F,
        0x202A, 0x202B, 0x202C, 0x202D, 0x202E,
        0x2066, 0x2067, 0x2068, 0x2069
    ]

    static func sanitize(_ value: String) -> String {
        var result = ""
        var pendingSpace = false
        for scalar in value.unicodeScalars {
            let disallowed = CharacterSet.controlCharacters.contains(scalar)
                || CharacterSet.newlines.contains(scalar)
                || bidirectionalControls.contains(scalar.value)
            if disallowed || CharacterSet.whitespaces.contains(scalar) {
                pendingSpace = !result.isEmpty
                continue
            }
            if pendingSpace {
                result.append(" ")
                pendingSpace = false
            }
            result.append(String(scalar))
        }
        let normalized = result.trimmingCharacters(in: .whitespacesAndNewlines)
        return normalized.isEmpty ? "予定名なし" : normalized.prefixingUnicodeScalars(80)
    }
}

@MainActor
final class TodayFocusTextAdapter {
    private let broker: CapabilityBroker
    private let activationLease: PocketAppActivationLease?

    init(
        broker: CapabilityBroker,
        activationLease: PocketAppActivationLease? = nil
    ) {
        self.broker = broker
        self.activationLease = activationLease
    }

    func listToday(
        timezone: TimeZone,
        principal: CapabilityPrincipal,
        permissions: CapabilityPermissionSet,
        now: Date = Date()
    ) async throws -> [TodayFocusCalendarEvent] {
        try activationLease?.requireActive()
        let nonce = UUID().uuidString.lowercased()
        let plan = CapabilityExecutionPlan(
            id: "today-focus-read:\(nonce)",
            createdAt: now,
            origin: .text,
            principal: principal,
            appContext: nil,
            steps: [
                CapabilityPlanStep(
                    id: "listCalendar",
                    capability: PocketCapabilityKeys.calendarList,
                    arguments: [
                        "range": .string("today"),
                        "timezone": .string(timezone.identifier)
                    ],
                    idempotencyKey: "today-focus-read.\(nonce)",
                    dependencies: []
                )
            ],
            requiredPermissions: ["calendar.events.read"]
        )
        let preparation = try broker.prepare(plan, permissions: permissions, now: now)
        guard preparation.approvalRequest == nil else {
            throw CapabilityBrokerError.invalidPlan("read_approval")
        }
        let receipt = try await executeTracked {
            try await self.broker.execute(
                plan,
                permissions: permissions,
                approvalGrant: nil,
                now: now
            )
        }
        try activationLease?.requireActive()
        guard receipt.status == .succeeded,
              let output = receipt.steps.first?.output,
              case .array(let events)? = output["events"] else {
            throw CapabilityBrokerError.unavailable(PocketCapabilityKeys.calendarList)
        }
        return try events.map { value in
            guard case .object(let event) = value,
                  case .string(let eventRef)? = event["eventRef"],
                  case .string(let safeTitle)? = event["safeTitle"],
                  case .string(let startValue)? = event["start"],
                  case .string(let endValue)? = event["end"],
                  let start = CapabilityDateCodec.date(from: startValue),
                  let end = CapabilityDateCodec.date(from: endValue) else {
                throw CapabilityBrokerError.unavailable(PocketCapabilityKeys.calendarList)
            }
            return TodayFocusCalendarEvent(
                eventRef: eventRef,
                safeTitle: safeTitle,
                start: start,
                end: end
            )
        }
    }

    func prepareFocus(
        event: TodayFocusCalendarEvent,
        durationSeconds: Int,
        purpose: String,
        principal: CapabilityPrincipal,
        permissions: CapabilityPermissionSet,
        now: Date = Date(),
        timeZone: TimeZone = .current
    ) throws -> TodayFocusDraft {
        try activationLease?.requireActive()
        guard (1...86_400).contains(durationSeconds),
              !purpose.isEmpty,
              purpose.unicodeScalars.count <= 10_000 else {
            throw CapabilityBrokerError.invalidPlan("today_focus_input")
        }
        let nonce = UUID().uuidString.lowercased()
        let stableDate = Self.dateKey(now, timeZone: timeZone)
        let approvalText = TodayFocusApprovalText.sanitize(purpose)
        let plan = CapabilityExecutionPlan(
            id: "today-focus-write:\(nonce)",
            createdAt: now,
            origin: .text,
            principal: principal,
            appContext: nil,
            steps: [
                CapabilityPlanStep(
                    id: "startTimer",
                    capability: PocketCapabilityKeys.timerStart,
                    arguments: [
                        "durationSeconds": .integer(durationSeconds),
                        "title": .string(approvalText),
                        "sourceRef": .string(event.eventRef)
                    ],
                    idempotencyKey: "today-focus-timer.\(nonce)",
                    dependencies: []
                ),
                CapabilityPlanStep(
                    id: "savePurpose",
                    capability: PocketCapabilityKeys.stickyUpsert,
                    arguments: [
                        "stableKey": .string("today-focus:\(stableDate)"),
                        "title": .string("今日の目的"),
                        "body": .string(approvalText),
                        "color": .string("yellow")
                    ],
                    idempotencyKey: "today-focus-sticky.\(nonce)",
                    dependencies: ["startTimer"]
                )
            ],
            requiredPermissions: ["sticky.write", "timer.write"]
        )
        return TodayFocusDraft(
            plan: plan,
            preparation: try broker.prepare(plan, permissions: permissions, now: now),
            approvalText: approvalText
        )
    }

    func approveAndExecute(
        _ draft: TodayFocusDraft,
        permissions: CapabilityPermissionSet,
        now: Date = Date()
    ) async throws -> CapabilityWorkflowReceipt {
        try activationLease?.requireActive()
        guard let request = draft.preparation.approvalRequest else {
            throw CapabilityBrokerError.approvalRequired
        }
        let grant = try broker.decideApproval(
            requestID: request.id,
            planDigest: draft.preparation.planDigest,
            decision: .approve,
            now: now
        )
        let receipt = try await executeTracked {
            try await self.broker.execute(
                draft.plan,
                permissions: permissions,
                approvalGrant: grant,
                now: now
            )
        }
        try activationLease?.requireActive()
        return receipt
    }

    func reject(
        requestID: String,
        planDigest: String,
        now: Date
    ) throws -> CapabilityApprovalGrant {
        try activationLease?.requireActive()
        return try broker.decideApproval(
            requestID: requestID,
            planDigest: planDigest,
            decision: .reject,
            now: now
        )
    }

    private func executeTracked<T: Sendable>(
        _ operation: @escaping @MainActor @Sendable () async throws -> T
    ) async throws -> T {
        try activationLease?.requireActive()
        let task = Task { @MainActor in
            try Task.checkCancellation()
            return try await operation()
        }
        let registration = activationLease?.registerCancellation { task.cancel() }
        defer { activationLease?.unregisterCancellation(registration) }
        return try await withTaskCancellationHandler {
            try await task.value
        } onCancel: {
            task.cancel()
        }
    }

    private static func dateKey(_ date: Date, timeZone: TimeZone) -> String {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = timeZone
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }
}
